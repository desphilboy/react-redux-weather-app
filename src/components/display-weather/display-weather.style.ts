import styled from "styled-components";

export const WeatherRowStyle = styled.div`
	border: 1px solid orange;
	border-radius: 10px;
	margin: 2px;
	display: flex;
	max-width: 200px;
	flex-direction: column;
	.city-country {
		font-weight: bold;
		font-size: 20px;
	}
	.date-time {
		font-size: 10px;
		font-weight: bold;
	}
	.main-weather {
		div.maintext {
			color: Orange;
			font-size: 24px;
			font-weight: bold;
			margin: 0;
		}

		div.description {
		}
		display: flex;
		flex-direction: column;
		justify-contents: center;

		img {
			margin-top: -40px;
			margin-bottom: -30px;
		}
	}
	.temperature {
		margin-top: 24px;
		display: flex;
		flex-direction: row;
		justify-content: center;
		img {
			width: 60px;
			height: 60px;
			margin-top: -20px;
			margin-bottom: -5px;
		}
		div.maintext {
			color: Blue;
			font-size: 28px;
			font-weight: bold;
			margin: 0;
		}
	}
	.feels-like {
		margin-top: 10px;
		font-size: 20px;
		color: darkGray;
		margin-bottom: 30px;
	}
`;

export const DisplayWeatherStyle = styled.div`
	flex-wrap: wrap;
	width: 80%;
	display: flex;
	flex-direction: row;
`;
