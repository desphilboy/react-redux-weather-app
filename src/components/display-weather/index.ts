import DisplayWeather from "./display-weather";

export default DisplayWeather;
