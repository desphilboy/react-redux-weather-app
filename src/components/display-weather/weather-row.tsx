import React from "react";
import moment from "moment-timezone";
import { WeatherRowStyle } from "./display-weather.style";
import temperatureIcon from "../../icons/svg/HeatIcon.svg";

export const WeatherRow = ({ weather }) => {
	const { icon, main, description } = weather.data.weather[0];
	const { temp, feels_like , pressure,humidity, temp_min, temp_max} = weather.data.main;
	const dateTime = Date.now();
	
	return (
		<WeatherRowStyle>
			<div className="city-country">{`${weather.data.name}, ${weather.data.sys.country}`}</div>
			<div className="main-weather">
				<div className="maintext">{main}</div>
				<img
					alt={main}
					src={`http://openweathermap.org/img/wn/${icon}@4x.png`}
				/>
				<div>{description} </div>
			</div>
			<div className="date-time">
				{moment
					.tz(dateTime, "Australia/Melbourne")
					.format("YYYY/MM/DD hh:mm:ss")}
			</div>

			<div className="temperature">
				<img alt="temperature" src={temperatureIcon} />
				<div className="maintext">{temp}</div>
			</div>
			<div className="feels-like">Feels like: {feels_like}</div>

			<div className="pressure"> pressure: {pressure}</div>
			<div className="hamadity">humidity: {humidity}</div>
			<div className= "temp_min">minimum temperature: {temp_min}</div>
		</WeatherRowStyle>
	);
};

export default WeatherRow;
