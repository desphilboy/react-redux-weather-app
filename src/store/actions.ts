import { createActions } from "redux-actions";

export const START_API = "START_API";
export const ERROR_API = "ERROR_API";
export const DONE_API = "DONE_API";

export const SET_COUNTRY = "SET_COUNTRY";

export const SET_UNITS = "SET_UNITS";

export const CLEAR_RECORDS = "CLEAR_RECORDS";

export const { startApi, doneApi, errorApi } = createActions({
	START_API: () => ({ data: false, inProgress: true, error: false }), // payload creator
	DONE_API: (data) => ({ data, inProgress: false, error: false }),
	ERROR_API: (error) => ({ data: false, inProgress: false, error }),
});

export const setUnits = (unitsType) => ({
	type: SET_UNITS,
	payload: unitsType,
});

export const setCountry = (country) => ({
	type: SET_COUNTRY,
	payload: country,
});

export const clearRecords = () => ({
	type: CLEAR_RECORDS,
});
