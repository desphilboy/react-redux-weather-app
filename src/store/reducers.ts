import {
   START_API,
   ERROR_API,
   DONE_API,
   SET_COUNTRY,
   SET_UNITS,
   CLEAR_RECORDS,
} from "./actions";

export const INITIAL_STATE = {
   country: "AU",
   units: "metric",

   apiCall: { data: false, error: false, inProgress: false },
   weatherRecords: [],
};

export const apiCall = (state = INITIAL_STATE.apiCall, action) => {
   switch (action.type) {
      case START_API:
      case ERROR_API:
      case DONE_API:
         return { ...state, ...action.payload };
      default:
         return state;
   }
};

export const country = (state = INITIAL_STATE.country, action) => {
   switch (action.type) {
      case SET_COUNTRY:
         return action.payload;
      default:
         return state;
   }
};

export const units = (state = INITIAL_STATE.units, action) => {
   switch (action.type) {
      case SET_UNITS:
         return action.payload;
      default:
         return state;
   }
};

export const weatherRecords = (
   state = INITIAL_STATE.weatherRecords,
   action
) => {
   switch (action.type) {
      case DONE_API:
         return [...state, action.payload.data];
      case CLEAR_RECORDS:
         return [];
      default:
         return state;
   }
};
