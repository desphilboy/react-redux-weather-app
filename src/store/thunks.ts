import { startApi, doneApi, errorApi } from "./actions";

const appId = "ce242e6f1909e3ca32395ead09817fb6";
const baseUrl = "http://api.openweathermap.org/data/2.5/weather";

export const fetchWeather = ({ city = "Melbourne" } = {}) => (
	dispatch,
	getState
) => {
	const { country, units } = getState();

	let fetchurl = `${baseUrl}?q=${city || "Melbourne"},${
		country || "AU"
	}&units=${units || "metric"}&appid=${appId}`;

	dispatch(startApi());
	fetch(fetchurl)
		.then((response) => response.json())
		.then((data) => {
			console.log("data in fetchWeather", data);
			dispatch(doneApi(data));
		})
		.catch((err) => {
			console.log(err);
			dispatch(errorApi(err));
		});
};
