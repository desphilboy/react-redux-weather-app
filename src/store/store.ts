import { createStore, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { apiCall, country, units } from "./reducers";

const reducer = combineReducers({ apiCall, country, units });

const store = createStore(reducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;
