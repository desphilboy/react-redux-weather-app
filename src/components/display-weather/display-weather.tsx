import React, { useEffect } from "react";
import { connect } from "react-redux";
import { DisplayWeatherStyle } from "./display-weather.style";
import { clearRecords, setCountry } from "../../store/actions";
import { fetchWeather } from "../../store/thunks";

import WeatherRow from "./weather-row";

interface DisplayViewInterface {
	rows: object[];
	clearAll: () => void;
	resetCountry: () => void;
	fetchMajorCities: () => void;
}

export const DisplayView: React.FunctionComponent<DisplayViewInterface> = ({
	rows = [],
	clearAll,
	resetCountry,
	fetchMajorCities,
}) => {
	useEffect(() => {
		if (rows.length === 0) {
			resetCountry();
			fetchMajorCities();
		}
	});
	return (
		<DisplayWeatherStyle>
			{rows.length ? (
				<button onClick={clearAll}>X</button>
			) : (
				<div>Select a city in above box to see the weather</div>
			)}
			{rows.map((row, index) => (
				<WeatherRow weather={row} key={`key-${index}`} />
			))}
		</DisplayWeatherStyle>
	);
};

export const mapDispatch = (dispatch) => ({
	clearAll: () => dispatch(clearRecords()),
	resetCountry: () => dispatch(setCountry("AU")),
	fetchMajorCities: () => {
		dispatch(fetchWeather({ city: "Melbourne" }));
		dispatch(fetchWeather({ city: "Sydney" }));
		dispatch(fetchWeather({ city: "Brisbane" }));
		dispatch(fetchWeather({ city: "Adelaide" }));
	},
});

export const mapState = (state) => ({
	rows: state.weatherRecords,
});

export default connect(mapState, mapDispatch)(DisplayView);
