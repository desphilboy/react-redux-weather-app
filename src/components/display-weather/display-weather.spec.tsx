import React from "react";
import { mapDispatch, mapState, DisplayView } from "./display-weather";
import * as THUNK from "../../store/thunks";
import testRenderer from "react-test-renderer";
import { SAMPLEROWS } from "./__fixtures__/weather-data.sample";
describe("displayWeather", () => {
	describe("mapDispatch", () => {
		const dispatch = jest.fn();
		let map;
		beforeEach(() => {
			map = mapDispatch(dispatch);
			jest.clearAllMocks();
		});

		it("returns clearAll, resetCountry, fetchMajorCities", () => {
			expect(map).toEqual({
				clearAll: expect.any(Function),
				resetCountry: expect.any(Function),
				fetchMajorCities: expect.any(Function),
			});
		});

		it("clearAll, calls Dispatch with CLEAR_RECORDS action", () => {
			map.clearAll();
			expect(dispatch).toHaveBeenCalledWith({
				type: "CLEAR_RECORDS",
			});
		});

		it("fetchMajorcities, calls dispatch with fetchWeather for 4 major cities", () => {
			THUNK.fetchWeather = jest.fn((param) => param);

			map.fetchMajorCities();

			[ "Adelaide", "Brisbane"].forEach(
				(majorCity) => {
					expect(dispatch).toHaveBeenCalledWith({ city: majorCity });
				}
			);

			map.fetchMajorCities();
			[ "Adelaide", "Brisbane"].forEach(
				(majorCity) => {
					expect(THUNK.fetchWeather).toHaveBeenCalledWith({
						city: majorCity,
					});
				}
			);
		});

		it("resetCountry, calls dispatch with SET_COUNTRY action and value 'AU'", () => {
			map.resetCountry();
			expect(dispatch).toHaveBeenCalledWith({
				payload: "AU",
				type: "SET_COUNTRY",
			});
		});
	});

	describe("mapState", () => {
		const mockState = { weatherRecords: [{ somekey: "someValue" }] };
		let map;
		beforeEach(() => {
			map = mapState(mockState);
			jest.clearAllMocks();
		});

		it("uses state to fill rows value with the weatherRecords from state", () => {
			expect(map.rows).toEqual([
				{
					somekey: "someValue",
				},
			]);
		});
	});

	describe("View", () => {
		let testParams = {
			rows: [],
			clearAll: jest.fn(),
			resetCountry: jest.fn(),
			fetchMajorCities: jest.fn(),
		};

		it("renders correctly with empty rows", () => {
			const wrapper = testRenderer.create(
				<DisplayView {...testParams} />
			);
			expect(wrapper).toMatchSnapshot();
		});

		it("renders correctly with rows", () => {
			testParams.rows = SAMPLEROWS;
			const originalDate = global.Date.now;
			global.Date.now = () => 1601235380;
			const wrapper = testRenderer.create(
				<DisplayView {...testParams} />
			);
			expect(wrapper).toMatchSnapshot();
			global.Date.now = originalDate;
		});
	});
});
