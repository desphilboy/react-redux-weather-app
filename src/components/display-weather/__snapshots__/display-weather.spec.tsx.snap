// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayWeather View renders correctly with empty rows 1`] = `
<div
  className="sc-gtssRu eTfTVI"
>
  <div>
    Select a city in above box to see the weather
  </div>
</div>
`;

exports[`displayWeather View renders correctly with rows 1`] = `
<div
  className="sc-gtssRu eTfTVI"
>
  <button
    onClick={[MockFunction]}
  >
    X
  </button>
  <div
    className="sc-bdnylx gQgnDU"
  >
    <div
      className="city-country"
    >
      Sydney, AU
    </div>
    <div
      className="main-weather"
    >
      <div
        className="maintext"
      >
        Clear
      </div>
      <img
        alt="Clear"
        src="http://openweathermap.org/img/wn/01d@4x.png"
      />
      <div>
        clear sky
         
      </div>
    </div>
    <div
      className="date-time"
    >
      1970/01/19 10:47:15
    </div>
    <div
      className="temperature"
    >
      <img
        alt="temperature"
        src="HeatIcon.svg"
      />
      <div
        className="maintext"
      >
        15.85
      </div>
    </div>
    <div
      className="feels-like"
    >
      Feels like: 
      9.66
    </div>
    <div
      className="pressure"
    >
       pressure: 
      1024
    </div>
    <div
      className="hamadity"
    >
      humidity: 
      48
    </div>
    <div
      className="temp_min"
    >
      minimum temperature: 
      15
    </div>
  </div>
  <div
    className="sc-bdnylx gQgnDU"
  >
    <div
      className="city-country"
    >
      Brisbane, AU
    </div>
    <div
      className="main-weather"
    >
      <div
        className="maintext"
      >
        Clouds
      </div>
      <img
        alt="Clouds"
        src="http://openweathermap.org/img/wn/03d@4x.png"
      />
      <div>
        scattered clouds
         
      </div>
    </div>
    <div
      className="date-time"
    >
      1970/01/19 10:47:15
    </div>
    <div
      className="temperature"
    >
      <img
        alt="temperature"
        src="HeatIcon.svg"
      />
      <div
        className="maintext"
      >
        18.58
      </div>
    </div>
    <div
      className="feels-like"
    >
      Feels like: 
      15.8
    </div>
    <div
      className="pressure"
    >
       pressure: 
      1022
    </div>
    <div
      className="hamadity"
    >
      humidity: 
      63
    </div>
    <div
      className="temp_min"
    >
      minimum temperature: 
      17.78
    </div>
  </div>
  <div
    className="sc-bdnylx gQgnDU"
  >
    <div
      className="city-country"
    >
      Melbourne, AU
    </div>
    <div
      className="main-weather"
    >
      <div
        className="maintext"
      >
        Clear
      </div>
      <img
        alt="Clear"
        src="http://openweathermap.org/img/wn/01d@4x.png"
      />
      <div>
        clear sky
         
      </div>
    </div>
    <div
      className="date-time"
    >
      1970/01/19 10:47:15
    </div>
    <div
      className="temperature"
    >
      <img
        alt="temperature"
        src="HeatIcon.svg"
      />
      <div
        className="maintext"
      >
        13.05
      </div>
    </div>
    <div
      className="feels-like"
    >
      Feels like: 
      8.55
    </div>
    <div
      className="pressure"
    >
       pressure: 
      1025
    </div>
    <div
      className="hamadity"
    >
      humidity: 
      62
    </div>
    <div
      className="temp_min"
    >
      minimum temperature: 
      11.67
    </div>
  </div>
  <div
    className="sc-bdnylx gQgnDU"
  >
    <div
      className="city-country"
    >
      Adelaide, AU
    </div>
    <div
      className="main-weather"
    >
      <div
        className="maintext"
      >
        Clouds
      </div>
      <img
        alt="Clouds"
        src="http://openweathermap.org/img/wn/04d@4x.png"
      />
      <div>
        broken clouds
         
      </div>
    </div>
    <div
      className="date-time"
    >
      1970/01/19 10:47:15
    </div>
    <div
      className="temperature"
    >
      <img
        alt="temperature"
        src="HeatIcon.svg"
      />
      <div
        className="maintext"
      >
        20.55
      </div>
    </div>
    <div
      className="feels-like"
    >
      Feels like: 
      13.58
    </div>
    <div
      className="pressure"
    >
       pressure: 
      1019
    </div>
    <div
      className="hamadity"
    >
      humidity: 
      26
    </div>
    <div
      className="temp_min"
    >
      minimum temperature: 
      20
    </div>
  </div>
</div>
`;
